#! /bin/sh /usr/share/dpatch/dpatch-run
## 85_fix_dbescaping.dpatch by Michael Friedrich
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: This patch fixes escaping for command definitions 
## DP: Backported from git: 0a98462f01f6cdf7b4fda3ad0b78e0c8658f5761

@DPATCH@
X-Git-Url: https://git.icinga.org/?p=icinga-core.git;a=blobdiff_plain;f=module%2Fidoutils%2Fsrc%2Fdb.c;h=439bf7b5ede52f4c75f80aba728ca87b47b47dd2;hp=d542cff5cb105747ce0fac99dd01548892c961b7;hb=0a98462f01f6cdf7b4fda3ad0b78e0c8658f5761;hpb=32bd11ed15070a9ee2930de4d23c680ce4d28e0d;js=1

diff --git a/module/idoutils/src/db.c b/module/idoutils/src/db.c
index d542cff..439bf7b 100644
--- a/module/idoutils/src/db.c
+++ b/module/idoutils/src/db.c
@@ -1813,11 +1813,57 @@ char *ndo2db_db_escape_string(ndo2db_idi *idi, char *buf) {
 
 	z = strlen(buf);
 
+	/* escape characters */
+#ifndef USE_ORACLE /* everything else will be libdbi */
+	/* allocate space for the new string */
+
+        if ((newbuf = (char *) malloc((z * 2) + 1)) == NULL)
+                return NULL;
+
+        for (x = 0, y = 0; x < z; x++) {
+
+                if(idi->dbinfo.server_type==NDO2DB_DBSERVER_MYSQL){
+
+                         if(buf[x]=='\'' || buf[x]=='\"' || buf[x]=='*' || buf[x]=='\\' || buf[x]=='$' || buf[x]=='?' || buf[x]=='.' || buf[x]=='^' || buf[x]=='+' || buf[x]=='[' || buf[x]==']' || buf[x]=='(' || buf[x]==')')
+                                newbuf[y++]='\\';
+                }
+                else if(idi->dbinfo.server_type==NDO2DB_DBSERVER_PGSQL){
+
+			if (buf[x] == '\'' || buf[x] == '[' || buf[x] == ']' || buf[x] == '(' || buf[x] == ')')
+				newbuf[y++] = '\\';
+
+                	/* should be fixed with binding values */
+			/* if(buf[x]=='\'' )
+                               newbuf[y++]='\''; */
+		}
+		else {
+
+                       if(buf[x]=='\'' )
+                               newbuf[y++]='\'';
+
+		}
+
+                newbuf[y++] = buf[x];
+        }
+
+        /* terminate escape string */
+        newbuf[y] = '\0';
+
+        ndo2db_log_debug_info(NDO2DB_DEBUGL_PROCESSINFO, 2, "ndo2db_db_escape_string(%s) end\n", newbuf);
+        return newbuf;
+
+        //size_t res = dbi_conn_quote_string(idi->dbinfo.dbi_conn, &buf);
+ 
+        //ndo2db_log_debug_info(NDO2DB_DEBUGL_PROCESSINFO, 2, "ndo2db_db_escape_string(%s) end\n", buf);
+        //return buf;
+
+#else /* Oracle ocilib specific */
+
 	/* allocate space for the new string */
 	if ((newbuf = (char *) malloc((z * 2) + 1)) == NULL)
 		return NULL;
 
-	/* escape characters */
+
 	for (x = 0, y = 0; x < z; x++) {
 
                 if(buf[x]=='\'' )
@@ -1825,6 +1871,7 @@ char *ndo2db_db_escape_string(ndo2db_idi *idi, char *buf) {
 
 		newbuf[y++] = buf[x];
 	}
+#endif /* Oracle ocilib specific */
 
 	/* terminate escape string */
 	newbuf[y] = '\0';
